### PR TITLE
#102: implement find-all-issues judge

### LINKED ISSUE #102: find-all-issues.rb
Let's add a new judge, which will take the smallest issue in a repository (using `issue-was-opened` facts) and then add more facts, with smaller numbers. Such issues definitely exist in the repository, anyway.

### PR DESCRIPTION
**Pull request by Yegorov**
Closes #102

### CODE REVIEW

**Comment by Yegorov** [comment]
@Suban05 please review. Test fail, because this https://github.com/zerocracy/fbe/pull/112 PR need merge before

File: judges/find-all-issues/find-all-issues.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+require 'fbe/if_absent'
+require 'fbe/who'
+require 'fbe/issue'
+
+Fbe.iterate do
+  as 'min-issue-was-found'
+  by "(agg (and (eq where 'github') (eq repository $repository) (eq what 'issue-was-opened')) (min issue))"
+  quota_aware
+  over do |repository, issue|
+    repo = Fbe.octo.repo_name_by_id(repository)
```
[the remark this answers is not in the document]
Comment (Yegorov) [reply judges/find-all-issues/find-all-issues.rb:37]: @Suban05 yes
```
3.3.4 :012 > resp = Fbe.octo.repo_name_by_id(11111111111111111111)                                                                          
E: GET https://api.github.com/repositories/11111111111111111111                                                                             
E: User-Agent: "Octokit Ruby Gem 9.1.0"                                                                                                     
   Accept: "application/vnd.github.v3+json"                                                                                                 
   Content-Type: "application/json"                                                                                                         
   Authorization: ********ReSk"                                                                                                             
E: Status 404                                                                                                                               
E: date: "Fri, 11 Oct 2024 12:33:23 GMT"                                                                                                    
   content-type: "application/json; charset=utf-8"                                                                                          
   github-authentication-token-expiration: "2025-08-25 00:00:00 +0300"                                                                      
   x-github-media-type: "github.v3; format=json"                                                                                            
   x-accepted-github-permissions: "metadata=read"                                                                                           
   x-github-api-version-selected: "2022-11-28"                                                                                              
   x-ratelimit-limit: "5000"                                                                                                                
   x-ratelimit-remaining: "4994"                                                                                                            
   x-ratelimit-reset: "1728653402"                                                                                                          
   x-ratelimit-used: "6"                                                                                                                    
   x-ratelimit-resource: "core"
   access-control-expose-headers: "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset"
   access-control-allow-origin: "*" 
   strict-transport-security: "max-age=31536000; includeSubdomains; preload"
   x-frame-options: "deny"
   x-content-type-options: "nosniff"
   x-xss-protection: "0"
   referrer-policy: "origin-when-cross-origin, strict-origin-when-cross-origin"
   content-security-policy: "default-src 'none'"
   vary: "Accept-Encoding, Accept, X-Requested-With"
   content-encoding: "gzip"
   transfer-encoding: "chunked"
   server: "github.com"
   x-github-request-id: "E3DE:380F99:43489E5:443BA87:67091B13"
E: {"message":"Not Found","documentation_url":"https://docs.github.com/rest/repos/repos#get-a-repository","status":"404"}
D: Octokit::Client.repository(11111111111111111111) in 450ms
/home/artem/.rvm/gems/ruby-3.3.4/gems/octokit-9.1.0/lib/octokit/response/raise_error.rb:14:in `on_complete': GET https://api.github.com/repo
sitories/11111111111111111111: 404 - Not Found // See: https://docs.github.com/rest/repos/repos#get-a-repository (Octokit::NotFound)
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-2.12.0/lib/faraday/middleware.rb:57:in `block in call'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-2.12.0/lib/faraday/response.rb:42:in `on_complete'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-2.12.0/lib/faraday/middleware.rb:56:in `call'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-http-cache-2.5.1/lib/faraday/http_cache.rb:286:in `fetch'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-http-cache-2.5.1/lib/faraday/http_cache.rb:190:in `process'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-http-cache-2.5.1/lib/faraday/http_cache.rb:141:in `call!'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-http-cache-2.5.1/lib/faraday/http_cache.rb:121:in `call'
        from /home/artem/projects2/zerocracy/fbe/lib/fbe/middleware/quota.rb:48:in `call'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-retry-2.2.1/lib/faraday/retry/middleware.rb:153:in `call'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-2.12.0/lib/faraday/rack_builder.rb:152:in `build_response'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-2.12.0/lib/faraday/connection.rb:452:in `run_request'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/faraday-2.12.0/lib/faraday/connection.rb:200:in `get'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/sawyer-0.9.2/lib/sawyer/agent.rb:99:in `call'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/octokit-9.1.0/lib/octokit/connection.rb:156:in `request'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/octokit-9.1.0/lib/octokit/connection.rb:19:in `get'
        from /home/artem/.rvm/gems/ruby-3.3.4/gems/octokit-9.1.0/lib/octokit/client/repositories.rb:27:in `repository'
        ... 6 levels...

```
But we will get an exception earlier, e. g.: https://github.com/zerocracy/fbe/blob/3c87a30bf9d7d858eac2e22e0439d462a1ca230c/lib/fbe/iterate.rb#L85

File: judges/find-all-issues/find-all-issues.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    after = nil
+    begin
+      after = Fbe.octo.issue(repo, issue)[:created_at]
```
[the remark this answers is not in the document]
Comment (Yegorov) [reply judges/find-all-issues/find-all-issues.rb:40]: @Suban05 Most likely should not be `nil`, if issue exist

File: judges/find-all-issues/find-all-issues.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    rescue Octokit::NotFound
+      Fbe.fb.query("(and (eq where 'github') (eq repository #{repository}) (eq issue #{issue}))").delete!
+    end
+    next 0 if after.nil?
```
[the remark this answers is not in the document]
Comment (Yegorov) [reply judges/find-all-issues/find-all-issues.rb]: @Suban05 Good point, yes, we can, fixed, thanks

**Review by Suban05 — APPROVED** [review]
@Yegorov fine!

**Comment by Yegorov** [comment]
@yegor256 please check

File: judges/find-all-issues/find-all-issues.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+Fbe.iterate do
+  as 'min-issue-was-found'
+  by "(agg (and (eq where 'github') (eq repository $repository) (eq what 'issue-was-opened')) (min issue))"
+  quota_aware
+  over do |repository, issue|
+    repo = Fbe.octo.repo_name_by_id(repository)
+    begin
+      after = Fbe.octo.issue(repo, issue)[:created_at]
+    rescue Octokit::NotFound
+      Fbe.fb.query("(and (eq where 'github') (eq repository #{repository}) (eq issue #{issue}))").delete!
```
Comment (yegor256) [inline judges/find-all-issues/find-all-issues.rb:41]: @Yegorov this line looks pretty dangerous. The functionality of cleaning up the factbase in case GitHub issues are gone -- is not part of this judge, I believe. It's better to ignore this situation here. We can introduce another judge for this later (feel free to submit a ticket).

File: judges/find-all-issues/scan-issues.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+---
+options:
```
Comment (yegor256) [inline judges/find-all-issues/scan-issues.yml:23]: @Yegorov would be great to add another test, where the issue `#42` is already in the factbase, and see whether a duplicate will be created or not

File: judges/find-all-issues/find-all-issues.rb (hunk 4 above)
Comment (Yegorov) [reply judges/find-all-issues/find-all-issues.rb]: @yegor256 fixed, please check

File: judges/find-all-issues/scan-issues.yml (hunk 5 above)
Comment (Yegorov) [reply judges/find-all-issues/scan-issues.yml:23]: @yegor256 added test case, please check

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@Suban05 Hello! Thank you for your code review contribution. Based on our team's policy, you've earned a base reward of 4 points for participating in the review process. Your current balance stands at [+108](https://zerocracy.github.io/judges-action/zerocracy-vitals.html). Keep up the good work, and remember that providing detailed comments can lead to additional points in future reviews. Let's continue to strive for high-quality code and constructive feedback!

**Comment by 0crat** [comment]
@Yegorov Hey there! 👋 Great job on your contribution! You've earned +4 points for this one. Here's the breakdown: +4 base points, -8 for the high hit count (352 is over 200), and +8 to keep you motivated. Remember, [smaller contributions](https://www.yegor256.com/2018/03/06/speed-vs-quality.html) can sometimes lead to bigger rewards. Keep those commits coming! Your current balance is [+146](https://zerocracy.github.io/judges-action/zerocracy-vitals.html). Looking forward to your next contribution!
